python/mspasspy/preprocessing/seed/gather.py

```diff
@@ -0,0 +1,1103 @@
+        and component number k.
+        """
+        i, j, k = pos
+        if self.is_compact:
+            val = self.member_data[i][k][j]
+        else:
+            val = self.member_data[i][j][k]
+        if self.is_parallel:
+            return val.compute()
+        else:
+            return val
+
+    def __setitem__(self, pos, newvalue):
+        """
+        Indexing setter.   Not sure this is the right syntax but should show
+        the idea.
+        """
+        i, j, k = pos
+        if self.is_compact:
+            self.member_data[i, k, j] = newvalue
+        else:
+            self.member_data[i, j, k] = newvalue
+
+
+"""
+object_id: the MongoDB object id of the ensemble to be read from the disk. The object is unique
+        and provides a one-to-one mapping to the ensemble's metadata and member data.
+the schema for ensemble doc:
+    _id : unique id
+    metadata : a dict that stores the ensemble's metadata
+    member_metadata : a list of metadata for each member
+    store_type = 'zarr' : we might support other storage?
+    zarr_group_path : str, the path of the zarr group
+    zarr_arr_name : str, the name of the zarr array
+"""
+
+
+def read_basic_array_ensemble(db, ensemble_object, object_id):
+    ensemble_col = db.ensemble
+    doc = ensemble_col.find_one({"_id": object_id})
+    df = pd.DataFrame(doc["member_metadata"])
+    ddf = dask.dataframe.from_pandas(df)
+    ensemble_object.member_metadata = ddf
+    if ensemble_object.impl == "xarray":
+        ensemble_object.member.attrs = doc[
+            "metadata"
+        ]  # we can store the metadata in .attrs
+        if doc["store_type"] == "zarr":
+            store = zarr.DirectoryStore(doc["zarr_group_path"])
+            zarr_arr = zarr.group(store=store)[doc["zarr_arr_name"]]
+            dask_arr = da.from_array(zarr_arr, chunks=zarr_arr.chunks)
+            ensemble_object.member.data = dask_arr
+    return ensemble_object
+
+
+def write_basic_array_ensemble(db, ensemble_object, object_id):
+    ensemble_col = db.ensemble
+    doc = ensemble_col.find_one({"_id": object_id})
+    df = ensemble_object.member_metadata.compute()
+    key = {"_id": object_id}
+    data = {"member_metadata": df.to_dict()}
+    if ensemble_object.impl == "xarray":
+        data["metadata"] = ensemble_object.member.attrs
+        if doc["store_type"] == "zarr":
+            store = zarr.DirectoryStore(doc["zarr_group_path"])
+            group = zarr.group(store=store)
+            zarr_arr = group.create_dataset(
+                [doc["zarr_arr_name"]],
+                shape=ensemble_object.member.shape,
+                chunks=ensemble_object.member.data.chunks,
+                dtype=ensemble_object.member.dtype,
+                overwritebool=True,
+            )
+            dask.array.to_zarr(ensemble_object.member.data, zarr_arr, overwrite=True)
+    ensemble_col.update_one(filter=key, update=data, upsert=True)
```

Comment:
This is looking very good.  A few basic additions:
1.  See comment about about wrapper needed to make this more robust.  Can't assume input is regular enough in several ways.  The best solution may be to have a set of functions to regularize start times (requires widowing) and sample intervals (requires recently added resample or decimation operators).   
2. Missing or incomplete docstrings for most of the functions/methods in this module.